torch/csrc/autograd/functions/accumulate_grad.cpp

```diff
@@ -6,80 +6,55 @@
 #include "torch/csrc/autograd/functions/utils.h"
 #include "torch/csrc/utils/auto_gpu.h"
 
+using at::Tensor;
+
 namespace torch { namespace autograd {
 
-AccumulateGrad::AccumulateGrad(std::shared_ptr<Variable> _variable)
-    : variable(_variable)
-    , variable_grad(_variable->grad) {
-  is_executable = _variable->requires_grad;
+AccumulateGrad::AccumulateGrad(Variable variable_)
+   : variable(std::move(variable_)) {
   num_inputs = 1;
-}
-
-auto AccumulateGrad::acc_inplace(std::shared_ptr<Variable>& grad,
-    std::shared_ptr<Variable>& new_grad) -> void {
-  auto& grad_data = grad->data;
-  auto& new_grad_data = new_grad->data;
-  AutoGPU guard(grad_data);
-
-  if (grad_data.type().isSparse() && !new_grad_data.type().isSparse()) {
-    grad->data = new_grad_data + grad_data;
-  } else {
-    grad_data += new_grad_data;
-  }
+  is_executable = 1;
 }
 
 auto AccumulateGrad::apply(const variable_list& grads) -> variable_list {
   // XXX: this method is not thread-safe!
   check_input_variables("AccumulateGrad", grads, 1, 0);
-  auto new_grad = grads[0];
-
-  if (!new_grad) return {};
-
-  auto var = variable.lock();
-  // It's possible that the Variable went out of scope and was freed.
-  // We still need to handle the unlikely case of someone holding to its grad.
-  if (!var) {
-    auto var_grad = variable_grad.lock();
-    // Everything was freed. Nothing to do.
-    if (!var_grad) return variable_list();
-    // Now here's the hard part. If both the new_grad and var_grad are volatile
-    // then we just acumulate the data in place (as we'd do if the Variable was
-    // alive). Otherwise, we'd need to perform the out-of-place reduction, but
-    // since the user only holds a reference to .grad and there's no way to
-    // give him the new Value, we just assume that they know these attributes
-    // are changing when using higher order graphs.
-    if (!var_grad->is_volatile || !new_grad->is_volatile) return variable_list();
-    acc_inplace(var_grad, new_grad);
-    return variable_list();
-  }
 
-  if (var->grad_fn)
+  if (!grads[0].defined())
+    return {};
+  if (!variable.defined())
+    throw std::logic_error("leaf variable was freed");
```

Comment:
So? Leaf Variables can be freed and it's not an error so let's not imply this. Mention that it has to do with the fact that it was still used in the graph.

Also, it's a condition that we handle now, so it's a breaking change. Not sure how important this is

Reply:
We can probably remove the check. AccumulateGrad has a strong pointer to `variable` so it won't get freed unless someone explicitly calls `AccumulateGrad.variable.reset()` or something similar.

Reply:
Yes, that sounds good. The check is a bit confusing, but a comment would be nice